ocs/ocs_agent.py

```diff
@@ -988,3 +1002,228 @@ def add_message(self, message, timestamp=None):
+            1629464204.781 Status is now "done".
+          other keys in .session: op_code, data
+
+    """
+    def __init__(self, params):
+        self._params = params
+        self._checked = set()
+
+    def get(self, key, default=ParamError(''), check=None, cast=None, type=None,
+            choices=None, treat_none_as_missing=True):
+        """Retrieve a value from the wrapped params dict, with optional type
+        checking, casting, and validity checks.  If a parameter is
+        found to be missing, or its value not valid, then a ParamError
+        is raised.
+
+        In Agent Op implementatations, the ParamError will be caught
+        by the API wrapper and automatically propagated to the caller.
+        The Operation session will be marked as "done", with
+        success=False.
+
+        This works best if the implementation validates all parameters
+        *before* beginning any Operation activities!
+
+        Args
+        ----
+        key : str
+          The name of the parameter to extract.
+        default : any
+          The value to use if the value is not set.  If this isn't
+          explicitly set, then a missing key causes an error to be
+          raised (see also the treat_none_as_missing arg).
+        check : callable
+          A function that will validate the argument; if the function
+          returns False then a ParamError will be raised.
+        cast : callable
+          A function to run on the value to convert it.  For example
+          cast=str.lower would help convert user argument "Voltage" to
+          value "voltage".
+        type : type
+          Class to which the result will be compared, unless it is
+          ``None``.  Note that if you pass ``type=float``, ``int``
+          values will automatically be cast to ``float`` and accepted
+          as valid.
+        choices : list
+          Acceptable values for the parameter.  This is checked after
+          casting.
+        treat_none_as_missing : bool
+          Determines whether a value of ``None`` for a parameter
+          should be treated in the same way as if the parameter were
+          not set at all.  See notes.
+
+        Returns
+        -------
+        The fully processed value.
+
+        Notes
+        -----
+
+        The default behavior is to treat ``{'param': None}`` as the
+        same as ``{}``; i.e. passing ``None`` as the value for a
+        parameter is the same as leaving the parameter unset.  In both
+        of these cases, unless a ``default=...`` is specified, a
+        ``ParamError`` will be raised.  Note this doesn't preclude you
+        from setting ``default=None``, which would effectively convert
+        ``{}`` to ``{'param': None}``.  If you really need to block
+        ``{}`` while allowing ``{'param': None}`` to be taken at face
+        value, then set ``treat_none_as_missing=False``.
```

Comment:
I don't quite understand why all of this is necessary. Does this mean you can't have a required parameter where `None` is a valid input? If that's the case then I think it would be worthwhile to figure out how to do this without requiring agent-writers to set `treat_none_as_missing`, since this seems like an easy-to-miss source of future runtime errors. Does treating None as missing benefit the user in some way I'm not getting?

Reply:
> I don't quite understand why all of this is necessary. Does this mean you can't have a required parameter where `None` is a valid input? If that's the case then I think it would be worthwhile to figure out how to do this without requiring agent-writers to set `treat_none_as_missing`, since this seems like an easy-to-miss source of future runtime errors. Does treating None as missing benefit the user in some way I'm not getting?

Short answer: I don't think this is that "easy-to-miss".  The param decorator will probably set up some checks, such as "type=float", and if the Agent code somewhere does "if param['x'] is None" then the SOCS police reviewing the PR will say "hey you clearly didn't test this with x=None".  Even if this does sneak through, I think this is an easier bug to identify and fix than the bugs that arise from using None for something other than triggering default behavior.

Long answer:

What this boils down to is my strong sense that in complex, layered systems, code written under the convention that param=None should trigger the default behavior (the same as not passing param) is more maintainable.  I guess I’m writing a blog post about this now.  Thanks in advance for reading.

Let’s first consider optional parameters, meaning they have a default value if not explicitly passed.  Suppose the default is 7.  Two possible function signatures for this are:
```
  def func_a(optional=None)
  def func_b(optional=7)
```
In the case of `func_a`, the body of the code has to catch `optional=None` and convert it to 7.  So the following have the same effect:
```
   func_a(optional=None)
   func_a()
   func_b(optional=7)
```
but this is different:
```
   func_b(optional=None)
```
The body for `func_a` needs to check for `None` and convert it to 7, but that’s quick.  Conversely the code for `func_b`, if you don't think we should reject None automatically, would need to reject `optional=None` … unless you want it to be significant; you want it to trigger some other behavior, that cannot be expressed by a numerical value for optional.  Fine.

So the situation where you need `func_b`-like behavior is when all of the following are true of the parameter `optional`:

- Numerical values are (mostly) accepted by the function.
- None is an acceptable value, and it triggers some behavior that can’t be reproduced by a particular numerical value (not even 0 or -1!).
- The desired default behavior corresponds to some numerical value, not None.

(I assume optional is a numerical parameter for clarity.  Note that with the param decorator, you can get the behavior described above with the arguments:“treat_none_as_missing=False, default=7”.)

But I actually think that situation above is to be avoided.  Because I have found that if instead one codes to the rule of thumb _“passing None to an optional parameter should trigger default behavior”_, then any code that uses your code, especially if it sort of wraps your code, is much easier to read and understand.  If you allow `None` to trigger non-default behavior, then you always have to watch out for hidden gotchas and inconsistencies.  Here is what I mean:

Consider a wrapping function that calls `func_b`:
```
  def tune_array_and_then_do_b(tune_optional=None, b_optional=None):
    tune_func(optional=tune_optional)
    func_b(optional=b_optional)
```
Wouldn’t you expect this function, if called with no arguments, `tune_array_and_then_do_b()`, to be equivalent to doing `tune_func()` and then `func_b()`?  Wrong, because `func_b()` is the same as `func_b(optional=7)`.  To get that propagation of "default" behavior, you'd have to sign it like this:
```
  def tune_array_and_then_do_b(tune_optional=None, b_optional=7):
    tune_func(optional=tune_optional)
    func_b(optional=b_optional)
```
And now someone reading this function needs to wonder:  "b_optional=7?  Didn't I change func_b last month so the default is 3?  Oops, I didn’t propagate that function’s new default value to … everywhere that tries to call the function with default parameters."

If one has everywhere followed the rule that `None` triggers default behavior, then this confusion doesn’t arise.

So… this sort of thing is why I have gradually developed a distaste for putting the defaults in the function signature in systems that I expect will eventually be layered.  And I tend to write functions that will eventually be layered.

There are some alternatives that give similar behavior as `func_b`.  Consider
```
   func_c(optional=None, do_optional_thing=None)
```
So the default behavior, might convert (None,None) to (optional=7, do_optional_thing=True).  And the user could prevent that by passing `do_optional_thing=False` (rather than `optional=None`).

Another possibility is to accept some other special value instead of `None`, to prevent the action.  I have sometimes used `False` to suppress a behavior that would otherwise expect a numerical parameter.  This makes type checking a bit more annoying, but it can still be done in the decorator and it’s not the worst lambda ever.  If you hate lambdas, and you extra hate lambdas in decorators, then we can define `float_or_false` and similar checking functions to help folks out and raise a good `ParamError`.

For the case of “mandatory” parameters – there is less of an issue here, since my diatribe above was all about defaults.  But in OCS we pass everything as keyword params, even mandatory ones that seem positional… so although this wrapping function seems less problematic than the ones above:
```
  def tune_array_and_then_do_b(tune_param, b_param):
    tune_func(mandatory=tune_param)
    func_b(mandatory=b_param)
```
… I think we’d be more likely to write things like this:
```
  def tune_array_and_then_do_b(tune_param=None, b_param=None):
    tune_func(mandatory=tune_param)
    func_b(mandatory=b_param)
```
and then you have the same uniformity issue: running `tune_array_and_then_do_b()` is not the same as running `tune_func()` and then `func_b()`.  (The former is valid, the latter will complain about missing arg.)


So… I think `None` triggering a non-default behavior should be discouraged.  I think using `False` to suppress something is probably better than using `None`, even in the case of “mandatory” arguments.  People that disagree are still free to set `treat_none_as_missing=False`.  But know that I will lie awake at night, brooding on the existence of such usage, terrified that somewhere in in our 9-layer observatory control code there could be a seemingly innocuous `=None` in a function signature that, you know, ruins everything in a really subtle way.

Reply:
I feel your pain and am not going to make you change it, I just wish there were a better solution in python... If we do adopt the practice of always using `None` as defaults and set the actual default in the body of the function then we miss out on being able to see what the defaults are in the signature, and we end up with a lot of
```
if optional is None:
    optional = 7
```
However writing it out those problems seem very insignificant and are outweighed by the benefits you so convincingly described above. 

Especially coming from pysmurf where there are layers on layers of function calls and hundreds of default arguments, some of which are inconsistent, I agree we should probably do everything we can to prevent that from happening.

Reply:
Thanks for reading.  Yes, one does pay a price for this choice.  In the agent at least... the defaults were never in the signature.  But they _will_ be in the decorator lines, and so the agent code won't have to "if optional is None: optional=default" all over the place.